core/commands/pin.go

```diff
@@ -65,118 +64,109 @@ var addPinCmd = &cmds.Command{
+			added, err := corerepo.Pin(n, api, req.Context, req.Arguments, recursive)
 			if err != nil {
-				res.SetError(err, cmdkit.ErrNormal)
-				return
+				return err
 			}
-			res.SetOutput(&AddPinOutput{Pins: cidsToStrings(added)})
-			return
+			return res.Emit(&AddPinOutput{Pins: cidsToStrings(added)})
 		}
 
 		out := make(chan interface{})
-		res.SetOutput((<-chan interface{})(out))
+
 		v := new(dag.ProgressTracker)
-		ctx := v.DeriveContext(req.Context())
+		ctx := v.DeriveContext(req.Context)
 
 		type pinResult struct {
 			pins []cid.Cid
 			err  error
 		}
 		ch := make(chan pinResult, 1)
 		go func() {
-			added, err := corerepo.Pin(n, api, ctx, req.Arguments(), recursive)
+			added, err := corerepo.Pin(n, api, ctx, req.Arguments, recursive)
 			ch <- pinResult{pins: added, err: err}
 		}()
 
-		ticker := time.NewTicker(500 * time.Millisecond)
-		defer ticker.Stop()
-		defer close(out)
-		for {
-			select {
-			case val := <-ch:
-				if val.err != nil {
-					res.SetError(val.err, cmdkit.ErrNormal)
-					return
-				}
+		errC := make(chan error)
+		go func() {
+			var err error
+			ticker := time.NewTicker(500 * time.Millisecond)
+			defer ticker.Stop()
+			defer func() { errC <- err }()
+			defer close(out)
 
-				if pv := v.Value(); pv != 0 {
+			for {
+				select {
+				case val := <-ch:
+					if val.err != nil {
+						err = val.err
+						return
+					}
+
+					if pv := v.Value(); pv != 0 {
+						out <- &AddPinOutput{Progress: v.Value()}
+					}
+					out <- &AddPinOutput{Pins: cidsToStrings(val.pins)}
+					return
+				case <-ticker.C:
 					out <- &AddPinOutput{Progress: v.Value()}
+				case <-ctx.Done():
+					log.Error(ctx.Err())
+					err = ctx.Err()
+					return
 				}
-				out <- &AddPinOutput{Pins: cidsToStrings(val.pins)}
-				return
-			case <-ticker.C:
-				out <- &AddPinOutput{Progress: v.Value()}
-			case <-ctx.Done():
-				log.Error(ctx.Err())
-				res.SetError(ctx.Err(), cmdkit.ErrNormal)
-				return
 			}
+		}()
+
+		err = res.Emit(out)
+		if err != nil {
+			return err
 		}
-	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			v, err := unwrapOutput(res.Output())
-			if err != nil {
-				return nil, err
-			}
 
+		return <-errC
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *AddPinOutput) error {
 			var added []string
 
-			switch out := v.(type) {
-			case *AddPinOutput:
-				if out.Pins != nil {
-					added = out.Pins
-				} else {
-					// this can only happen if the progress option is set
-					fmt.Fprintf(res.Stderr(), "Fetched/Processed %d nodes\r", out.Progress)
-				}
-
-				if res.Error() != nil {
-					return nil, res.Error()
-				}
-			default:
-				return nil, e.TypeErr(out, v)
+			if out.Pins != nil {
+				added = out.Pins
+			} else {
+				// this can only happen if the progress option is set
+				return fmt.Errorf("Fetched/Processed %d nodes\r", out.Progress)
```

Comment:
This is wrong.  This is not an error.

This command should likely use a PostRun.